### PR TITLE
Ee 70 product category

### PR DESCRIPTION
**Pull request by badrulme**

### CODE REVIEW

File: product-service/src/main/java/com/extremecoder/productservice/controller/CategoryController.java [hunk 1]
```diff
@@ -1,25 +1,39 @@
...
-    @GetMapping
-    public ResponseEntity<List<CategoryDto>> findAllCategories() {
-        return ResponseEntity.ok(categoryService.getAllCategories());
+    private final CategoryService service;
+
+
+    @PostMapping
+    public ResponseEntity<Response<Object, Object>> create(@Valid @RequestBody CategoryRegistrationRequest request) {
+
+        Response<Object, Object> response = Response.builder()
```
Comment (shshetudev) [inline product-service/src/main/java/com/extremecoder/productservice/controller/CategoryController.java]: - Please change <Object, Object>.
- Why we have to create a builder() here, it's already implemented in Response class.

File: product-service/src/main/java/com/extremecoder/productservice/controller/CategoryController.java [hunk 2]
```diff
@@ -1,25 +1,39 @@
...
 import org.springframework.web.bind.annotation.RestController;
 
-import java.util.List;
+import javax.validation.Valid;
+import java.util.Date;
 
 @RestController
 @RequestMapping("/api/v1/categories")
 @RequiredArgsConstructor
-@Slf4j
```
Comment (shshetudev) [inline product-service/src/main/java/com/extremecoder/productservice/controller/CategoryController.java:17]: Please add @Slf4j here. We are already using lombok. And logging is also needed for the controller layer.

File: product-service/src/main/java/com/extremecoder/productservice/dto/CategoryDetail.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    private Date createdDate;
+
+    private Date lastModifiedDate;
+
+    @Getter
+    @Setter
+    @Builder
+    @AllArgsConstructor
+    @NoArgsConstructor
+    public static class ParentCategory {
```
Comment (shshetudev) [inline product-service/src/main/java/com/extremecoder/productservice/dto/CategoryDetail.java]: Please separate it into a different files.

File: product-service/src/main/java/com/extremecoder/productservice/dto/CategoryRegistrationRequest.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+@Setter
+public class CategoryRegistrationRequest {
+
+    @NotBlank
+    @Length(max = 255)
+    private String name;
+    @Length(max = 500)
+    private String description;
+    @NotNull
+    private Boolean isParent;
```
Comment (shshetudev) [inline product-service/src/main/java/com/extremecoder/productservice/dto/CategoryRegistrationRequest.java:20]: Please change `Boolean` to boolean.

File: product-service/src/main/java/com/extremecoder/productservice/dto/Response.java [hunk 5]
```diff
@@ -1,25 +1,28 @@
...
+//    @JsonInclude(JsonInclude.Include.NON_NULL)
+//    private Long rowCount;
     @JsonInclude(JsonInclude.Include.NON_NULL)
-    private Integer numberOfElement;
-    @JsonInclude(JsonInclude.Include.NON_NULL)
-    private Long rowCount;
-    @JsonInclude(JsonInclude.Include.NON_NULL)
-    private List<ErrorResponseDto> errors;
+    private U errors;
+//    private List<FieldErrorResponse> errors;
```
Comment (shshetudev) [inline product-service/src/main/java/com/extremecoder/productservice/dto/Response.java]: Please remove commented code.

File: product-service/src/main/java/com/extremecoder/productservice/exception/GlobalExceptionHandler.java [hunk 6]
```diff
@@ -137,25 +139,50 @@ public ResponseEntity<Object> handleInternalException(Throwable ex) {
      * @return the response entity
      */
     @ExceptionHandler({BindException.class})
-    public ResponseEntity<Object> handleBindException(BindException ex) {
-        // todo: modify response
+    public ResponseEntity<Response<Object, Object>> handleBindException(BindException ex) {
```
Comment (shshetudev) [inline product-service/src/main/java/com/extremecoder/productservice/exception/GlobalExceptionHandler.java]: Please try to think something better.

File: product-service/src/main/java/com/extremecoder/productservice/mapper/CategoryMapper.java [hunk 7]
```diff
@@ -1,20 +1,16 @@
...
 @Mapper(unmappedTargetPolicy = ReportingPolicy.IGNORE)
 public interface CategoryMapper {
-    CategoryDto toDto(Category category);
-
-    Category toEntity(CategoryDto categoryDto);
-
-    List<CategoryDto> toDtos(List<Category> categories);
-
-    List<Category> toEntities(List<CategoryDto> categoryDtos);
+//    CategoryDto toDto(Category category);
```
Comment (shshetudev) [inline product-service/src/main/java/com/extremecoder/productservice/mapper/CategoryMapper.java]: Unused commented code.

File: product-service/src/main/java/com/extremecoder/productservice/utils/DataLoader.java [hunk 8]
```diff
@@ -24,17 +21,17 @@ public class DataLoader implements CommandLineRunner {
...
-        imageInfoRepository.save(fruitsImage);
-        ImageInformation bananaImage = new ImageInformation(null, 3L, "3");
-        imageInfoRepository.save(bananaImage);
-        Category fruitsVegetablesCat = new Category(null, "Fruits & Vegetables", fruitsVegetableImage, null);
-        categoryRepository.save(fruitsVegetablesCat);
-        Category fruitsCategory = new Category(null, "Fruits", fruitsImage, fruitsVegetablesCat);
-        categoryRepository.save(fruitsCategory);
-        Product product1 = new Product(null, "Banana", "Banana is the most popular fresh fruit in all over the world. It has lots of variety. ", bananaImage, fruitsCategory);
-        productRepository.save(product1);
+//        ImageInformation fruitsVegetableImage = new ImageInformation(null, 1L, "1");
```
Comment (shshetudev) [inline product-service/src/main/java/com/extremecoder/productservice/utils/DataLoader.java:24]: Unused commented code.

File: product-service/src/main/java/com/extremecoder/productservice/utils/ResponseBuilder.java [hunk 9]
```diff
@@ -56,20 +56,19 @@ public static Response getSuccessResponse(HttpStatus status, String message, Obj
...
-                .message(message)
-                .status(status.getReasonPhrase())
-                .statusCode(status.value())
-                .content(content)
-                .timeStamp(LocalDateTime.now())
-                .numberOfElement(numberOfElement)
-                .rowCount(rowCount)
-                .build();
-    }
+//    public static Response getSuccessResponse(HttpStatus status, String message, Object content, int numberOfElement, Long rowCount) {
```
Comment (shshetudev) [inline product-service/src/main/java/com/extremecoder/productservice/utils/ResponseBuilder.java]: Remove commented code.

File: product-service/src/main/java/com/extremecoder/productservice/controller/CategoryController.java (hunk 1 above)
Comment (badrulme) [reply product-service/src/main/java/com/extremecoder/productservice/controller/CategoryController.java]: - **Please change <Object, Object>**: What is your suggestion here?
- **Why we have to create a builder() here, it's already implemented in Response class:** Could not understand. Please explain a bit.

File: product-service/src/main/java/com/extremecoder/productservice/exception/GlobalExceptionHandler.java (hunk 6 above)
Comment (badrulme) [reply product-service/src/main/java/com/extremecoder/productservice/exception/GlobalExceptionHandler.java]: Please share your idea

**Pull request merged**